Change figure & figcaption accName computation

scottaohara opened the pull request:
resolves #325

This PR revises the mappings for `figure` and `figcaption` to remove the "labelled by" relations, changing them to "details for".
The accessible name computation for `figure` removes `figcaption`, a note is added to further reference this change.

- [x] need to update to UIA mapping (core aam issue: https://github.com/w3c/core-aam/issues/110)
- [x] do we want `figcaption` to map to `caption` (this is probably a 'yes', but want confirmation)? 
- [x] regardless of above, [`caption`](https://w3c.github.io/aria/#caption) will need to be updated in regards to its referencing `figure` and `figcaption` relationships (ARIA PR: https://github.com/w3c/aria/pull/1703)
- [x] [`aria-details`](https://w3c.github.io/aria/#aria-details) definition should be updated to reference this change (ARIA PR: https://github.com/w3c/aria/pull/1704)

Implementation commitment:

 * [ ] WebKit (https://bugs.webkit.org/show_bug.cgi?id=)
 * [x] Chromium 
 * [ ] Gecko (https://bugzilla.mozilla.org/show_bug.cgi?id=)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/html-aam/pull/359.html" title="Last updated on Dec 13, 2023, 1:35 PM UTC (2380d76)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/html-aam/359/ab38b66...2380d76.html" title="Last updated on Dec 13, 2023, 1:35 PM UTC (2380d76)">Diff</a>

scottaohara committed:
Change figure & figcaption accName computation

resolves #325

This PR revises the mappings for `figure` and `figcaption` to remove the "labelled by" relations, changing them to "details for".
The accessible name computation for `figure` removes `figcaption`, a note is added to further reference this change.

TODO: 
- [ ] need to check on updates to UIA mapping, if any.
- [ ] do we want `figcaption` to map to `caption`? 
- [ ] regardless of above, [`caption`](https://w3c.github.io/aria/#caption) will need to be updated in regards to its referencing `figure` and `figcaption` relationships
- [ ] [`aria-details`](https://w3c.github.io/aria/#aria-details) definition should be updated to reference this change

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara commented:
referencing this thread as a very good example of why this change needs to be made: https://github.com/WICG/proposals/issues/45

per what's being discussed there, if an entire video transcript were placed within a figcaption, the flattened text string of the transcript would become the accessible name for the parent figure.... and that would be awful.  In contrast, this is a perfectly reasonable thing to use a `figure` and `figcaption` for... hence these mapping changes would be beneficial here.

aleventhal commented on index.html:
```diff
@@ -1079,7 +1079,8 @@ <h3>HTML Element Role Mappings</h3>
                 <div class="relations">
-                  <span class="type">Relations:</span> `IA2_RELATION_LABEL_FOR` with parent <a href="#el-figure">`figure`</a> element
+                  <span class="type">Relations:</span> 
+                  `IA2_RELATION_DETAILS_FOR` with parent <a href="#el-figure">`figure`</a> element
```
The inverse relation, IA2_RELATION_DETAILS, should also be exposed.

aleventhal commented on index.html:
```diff
@@ -1096,7 +1097,7 @@ <h3>HTML Element Role Mappings</h3>
+                  `RELATION_DETAILS_FOR` with parent <a href="#el-figure">`figure`</a> element
                 </div>
                 <div class="ifaces">
                   <span class="type">Interfaces:</span> `AtkText`; `AtkHypertext`
```
Need to expose ATK_RELATION_DETAILS and ATK_RELATION_DETAILS_FOR.

cookiecrook reviewed (commented):
I'll have to stop short of "approving" since I'm not the right person to review ATK or IA2 changes. I see no issues with the other parts of the PR.

carmacleod commented:
Aside: As this will be one more element that can only reasonably get its accessible name from aria and not from html ([ignoring title attribute](https://twitter.com/codingchaos/status/1321859365368836097)), it becomes even more important to have a [new html mechanism to give an element an accessible name](https://github.com/whatwg/html/issues/7326).

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
resolve review comments

aleventhal commented on index.html:
```diff
@@ -1126,34 +1123,29 @@ <h3>HTML Element Role Mappings</h3>
-                  Accessible name derived from `figcaption` according to the <a href="#figure-element-accessible-name-computation">`figure` Element Accessible Name Computation</a>
+                <div class="property">
+                  <span class="type">Property:</span> 
+                  `DescribedBy`: points to the first child `figcaption`, if any
```
Check with dlibby from MS regarding the UIA mapping. They have added some support for aria-details — IIRC they mapped it to some UIA API, even though it wasn't a perfect fit.

aleventhal commented on index.html:
```diff
@@ -1126,34 +1123,29 @@ <h3>HTML Element Role Mappings</h3>
               </td>
-              <td class="comments"></td>
+              <td class="comments">
+                The first child instance of a <a href="#el-figcaption">`figcaption`</a> provides details 
```
What does it mean, "provides details for its parent". What's the exact AX API to use?
CC @cookiecrook

scottaohara replied:
the comment is not specific to ax, but a general this is what the element does.

but yes, would want to know what the ax mappings would be here.  Presently core aam lists `aria-details` as not mapped in ax api, so we need to know what those should be / if webkit is going to do this

aleventhal reviewed (approved):
Approve for ATK/IA2 parts, but I think we should follow up with Microsoft and Apple before merging.

aleventhal commented:
Ah, I found that Chrome uses NSAccessibilityDetailsElementsAttribute aka AXDetailsElements.
Here's the bug where we landed it: https://bugs.chromium.org/p/chromium/issues/detail?id=1216049
We followed WebKit's lead. There are links in the bug to the WebKit change as well.

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
indicate figcaption maps to caption

scottaohara committed:
add in AX details for figure's relationship to figcaption

cc @cookiecrook to make sure i got this right.  thank you

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

stevefaulkner commented:
@aleventhal it looks  this has been implemented in Chrome and as a result it has surfaced a bug in JAWS implementation of `aria-details` and more generally in the `IA2_RELATION_DETAILS` implementation. I have discussed with Glen Gordon and he is going to facilitate a fix in JAWS.

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara commented:
@benbeaudry could use your help with the UIA mappings so we can get this merged.

scottaohara commented:
Another thing that has come to my attention is adding a comment to NOT expose the details relationship in the event an author has associated the figure and figcaption via aria-labelledby.  In such a case, it would likely be less noisy to drop the details association since the author is specifically referencing the element as the figure's name.  e.g., 
```html
<figure aria-labelledby=n>
  <figcaption id=n>this becomes the name</figcaption>
  whatever the content of the figcaption is here.
</figure>
```

scottaohara committed:
finish up mapping detials

verified the UIA mappings with help from @benbeaudry 
made some of the wording consistent between mappings
added in a SHOULD to indicate that if a figure is explicitly provided an accessible name from its figcaption, then do not expose the details relationship, as that'd add redundancy.

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
attempt to fix build error

saying it doesn't know what accname is... so, replacing the instances of the links which may be causing the issue?

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
fix the broken accname/desc references

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara committed:
add comment regarding aria-details

see https://bugs.chromium.org/p/chromium/issues/detail?id=1426613#c4

aria-details with an empty string could be useful to authors if there was a specific need to _not_ expose a child figcaption - e.g., if the elements need to have remapped roles for whatever reason.

jnurthen commented:
@benbeaudry could use your help with the UIA mappings so we can get this merged.

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara commented:
@benbeaudry, @jnurthen this is ready for re-review and thanks to ben's core aam updates for aria-details i got the info i needed to complete this.  so, if everyone can give this one last look, then i think we can merge and get remaining implementation bugs filed for webkit/firefox.

cc @cookiecrook @jcsteh

cookiecrook commented on index.html:
```diff
@@ -2175,58 +2175,56 @@ <h4 id=el-figcaption>`figcaption`</h4>
         <div class="relations">
           <span class="type">Relations:</span>
-          `ATK_RELATION_LABEL_FOR` with parent <a href="#el-figure">`figure`</a> element
+          If the first instance of a child `figcaption` to its parent <a href="#el-figure">`figure`</a>, then `RELATION_DETAILS_FOR` with parent `figure` element.
```
I have changed this to what I hope is more understandable, but I'm not sure if I have inadvertently changed the meaning. Please double-check before merging the suggestion.

```suggestion
          If this is the first `figcaption` descendant in the related <a href="#el-figure">`figure`</a> element, then expose `RELATION_DETAILS_FOR` referencing the parent `figure` element.
```

cookiecrook commented on index.html:
```diff
@@ -2175,58 +2175,56 @@ <h4 id=el-figcaption>`figcaption`</h4>
     <tr>
       <th>Comments</th>
       <td>
+        <p>
+          Only the first child instance of a <a href="#el-figcaption">`figcaption`</a> this is present in the accessibility tree
+          provides details for its parent `figure`.</p>
```
Same caveat as above. Please ensure my editorial changes don't break your intended meaning.

```suggestion
        <p>Only the first descendant <a href="#el-figcaption">`figcaption`</a> element in a `figure` element provides details for its ancestor `figure` element.</p>
```

cookiecrook replied:
Aside from the editorial changes, why would only the first be related programmatically? Is this working around a bug in the API? If so, should that be filed?

scottaohara replied:
it's invalid HTML to have more than one figcaption within a figure, unless subsequent figcaptions are within their own nested figures. the figcaption must be a direct child of the figure.    i can see an argument to go beyond what html says is valid to provide more info for AT, but if we do that i'd want it to be in another PR.

so going to make some changes based on your editorial suggestions - but keeping references to 'child' and 'first instance'

cookiecrook commented on index.html:
```diff
@@ -2253,7 +2251,8 @@ <h4 id=el-figure>`figure`</h4>
         <div class="relations">
-          <span class="type">Relations:</span> `IA2_RELATION_LABELLED_BY` with child <a href="#el-figcaption">`figcaption`</a> element
+          <span class="type">Relations:</span> 
+          `IA2_RELATION_DETAILS` points to the first instance of a child `figcaption` if the element is present in the accessibility tree.
```
```suggestion
          `IA2_RELATION_DETAILS` points to the first instance of a descendant `figcaption` if the element is present in the accessibility tree of the ancestor `figure` element.
```

cookiecrook replied:
Should there be an exception for `aria-owns`?

cookiecrook commented on index.html:
```diff
@@ -2263,8 +2262,9 @@ <h4 id=el-figure>`figure`</h4>
         </div>
-        <div class="general">
-          Accessible name derived from `figcaption` according to the <a href="#figure-element-accessible-name-computation">`figure` Element Accessible Name Computation</a>
+        <div class="property">
+          <span class="type">Property:</span>
+          `DescribedBy` points to the first instance of a child `figcaption` if the element is present in the accessibility tree.
```
```suggestion
        <div class="property">
          <span class="type">Property:</span>
          `DescribedBy` references the first instance of a descendant `figcaption` element if the element is present in the accessibility tree of the ancestor `figure` element.
```

cookiecrook commented on index.html:
```diff
@@ -2278,7 +2278,8 @@ <h4 id=el-figure>`figure`</h4>
         <div class="relations">
-          <span class="type">Relations:</span> `ATK_RELATION_LABELLED_BY` with child <a href="#el-figcaption">`figcaption`</a> element
+          <span class="type">Relations:</span> 
+          `RELATION_DETAILS` points to the first instance of a child `figcaption` if the element is present in the accessibility tree.
```
```suggestion
          `RELATION_DETAILS` references the first instance of a descendant `figcaption` element if the element is present in the accessibility tree of the ancestor `figure` element.
```

cookiecrook commented on index.html:
```diff
@@ -2288,12 +2289,24 @@ <h4 id=el-figure>`figure`</h4>
         </div>
+        <div class="relations">
+          <span class="type">Property:</span>
+          `AXDetailsElements` points to the first child instance of a `figcaption` if the element is present in the accessibility tree.
```
```suggestion
          `AXDetailsElements` references the first instance of a descendant `figcaption` element if the element is present in the accessibility tree of the ancestor `figure` element.
```

cookiecrook commented on index.html:
```diff
@@ -2288,12 +2289,24 @@ <h4 id=el-figure>`figure`</h4>
       <th>Comments</th>
       <td>
+        <p>
+          If an author specifies a `figure` element to acquire its accessible name or description from the first instance
+          of a `figcaption` with an `aria-labelledby` or `aria-describedby` association, then user agents SHOULD NOT expose a 
+          details relationship between the `figure` and `figcaption` elements.</p>
```
```suggestion
          If an author references a `figure` element via `aria-labelledby` or `aria-describedby` on the first instance
          of a `figcaption` element, then user agents SHOULD NOT expose a 
          details relationship between the `figure` and `figcaption` elements.</p>
```

cookiecrook commented on index.html:
```diff
@@ -2288,12 +2289,24 @@ <h4 id=el-figure>`figure`</h4>
+          of a `figcaption` with an `aria-labelledby` or `aria-describedby` association, then user agents SHOULD NOT expose a 
+          details relationship between the `figure` and `figcaption` elements.</p>
+        <p>
+          An explicit declaration of the `aria-details` attribute on the `figure` element will override the implicit `aria-details` 
+          association with the element's child `figcaption` element, if one exists. If the `aria-details` attribute has an empty value,
+          then the `figure` element has no details relation to expose.</p>
```
This is complicated. Looking forward to it becoming testable someday.

cookiecrook commented on index.html:
```diff
@@ -16448,17 +16461,21 @@ <h4>`summary` Element Accessible Name Computation</h4>
       <section>
         <h4>`figure` Element Accessible Name Computation</h4>
+        <p>
+          A <a href="#el-figcaption">`figcaption`</a> provides additional details related to its parent 
```
```suggestion
          A <a href="#el-figcaption">`figcaption`</a> provides additional details related to its ancestor 
```

cookiecrook commented on index.html:
```diff
@@ -16448,17 +16461,21 @@ <h4>`summary` Element Accessible Name Computation</h4>
+          A <a href="#el-figcaption">`figcaption`</a> provides additional details related to its parent 
+          <a href="#el-figure">`figure`</a> element. A `figcaption` does not take part in the 
+          <a data-cite="accname-1.2/#dfn-accessible-name">accessible name</a> or <a data-cite="accname-1.2/#dfn-accessible-description">accessible description</a>
+          computation, unless explicitly referenced by an author. In such an instance, user agents SHOULD NOT expose the details relationship between the
+          `figure` `figcaption` elements, and instead expose the `figcaption` as the element's accessible name or description, as specified by the author.
```
"In such an instance" is unclear to me. I think this might be the intention:

```suggestion
          computation, unless explicitly referenced by an author. When an author uses `aria-labelledby` or `aria-describedby` to reference the first `figcaption` element from a`figure` element, user agents SHOULD NOT expose the details relationship between the
          `figure` and `figcaption` elements, and instead expose the `figcaption` as the element's accessible name or description, according to the <a href="https://w3c.github.io/accname/#mapping_additional_nd">AccName Name and Description Computation.</a>
```
Sorry I think you'll need to change that link above.

scottaohara replied:
i don't think we need to link to accName here, since the following naming steps do that already.  but i'm pulling in the other suggetions.  Let me know if you disagree and i should re-add the accName link again.

scottaohara committed:
update based on review suggestions

cc @cookiecrook

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

scottaohara commented:
created some markup tests with comments describing expectations:

```
<!--
  no accName
  no accDescription
  no details relationship due to lack of child figcaption element
-->
<figure>
  <img src="test.jpg" alt="test">
  <p>more content</p>
</figure>


<!--
  no accName
  no accDescription
  no details relationship due to lack of valid child figcaption element
-->
<figure>
  <img src="test.jpg" alt="test">
  <p>more content</p>
  <div>
    <figcaption>invalid details</figcaption>
  </div>
</figure>


<!--
  no accName
  no accDescription
  details relationship with child figcaption
-->
<figure>
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>


<!--
  no accName
  no accDescription
  details relationship with child figcaption
-->
<figure>
  <figcaption>details</figcaption>
  <img src="test.jpg" alt="test">
</figure>


<!--
  no accName
  no accDescription
  details relationship with first child figcaption
-->
<figure>
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
  <figcaption>not details</figcaption>
</figure>


<!--
  accName: 'label'
  no accDescription
  details relationship with child figcaption
-->
<figure aria-label=label>
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>


<!--
  accName: 'label'
  accDescription: 'title'
  details relationship with child figcaption
-->
<figure aria-label=label title=title>
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>


<!--
  accName: 'title'
  no accDescription
  details relationship with child figcaption
-->
<figure title=title>
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>


<!--
  accName: 'name'
  accDescription: 'title'
  details relationship with child figcaption
-->
<p id=n>name</p>
<figure aria-labelledby=n title=title>
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>


<!--
  accName: 'name'
  accDescription: 'title'
  details relationship with child figcaption
-->
<p id=n>name</p>
  <figure aria-labelledby=n aria-label=label title=title>
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>


<!--
  accName: 'caption'
  accDescription: 'title'
  no details relationship with child figcaption because it is used for accName
-->
<figure aria-labelledby=c aria-label=label title=title>
  <img src="test.jpg" alt="test">
  <figcaption id=c>caption</figcaption>
</figure>


<!--
  accName: 'label'
  accDescription: 'caption'
  no details relationship with child figcaption because it is used for accDescription
-->
<figure aria-label=label aria-describedby=c>
  <img src="test.jpg" alt="test">
  <figcaption id=c>caption</figcaption>
</figure>


<!--
  accName: 'label'
  accDescription: 'description'
  details relationship with child figcaption
-->
<figure aria-label=label aria-description="description">
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>


<!--
  no accName
  no accDescription
  no details relationship due to explicit aria-details=""
-->
<figure aria-details="">
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>


<!--
  accName: 'label'
  accDescription: 'description'
  details relationship with div#foo.  No details relationship with child figcaption
-->
<figure aria-details="foo">
  <img src="test.jpg" alt="test">
  <figcaption>not details</figcaption>
</figure>
<div id=foo>details</div>



<!-- 
  ###
  aria-owns test ideas
  ###
 -->

<!--
  no accName
  no accDescription
  details relationship with reparented figcaption via aria-owns
-->
<figure aria-owns=f>
  <img src="test.jpg" alt="test">
</figure>
<figcaption id=f>details</figcaption>


<!--
  no accName
  no accDescription
  details relationship with child figcaption. Reparented figcaption would be
  the last child element in the figcaption's a11y tree, thus it cannot be
  the first child instance used as a details reference.
-->
<figure aria-owns=f>
  <img src="test.jpg" alt="test">
  <figcaption>details</figcaption>
</figure>
<figcaption id=f>not details</figcaption>
```

cookiecrook commented:
Actually I didn't realize this update would require rolling back some WebKit implementation changes wrt figcaption... I think this needs WPT tests to fully qualify the implications of the change, especially the parts about accounting for multiple figcaptions... I think you could easily convert the above examples into a WPT test.

scottaohara committed:
Merge branch 'gh-pages' into figure-figcaption-revision

jcsteh commented on index.html:
```diff
@@ -2180,58 +2180,57 @@ <h4 id=el-figcaption>`figcaption`</h4>
       <th>Comments</th>
       <td>
+        <p>
+          Only the first child instance of a <a href="#el-figcaption">`figcaption`</a> element that is present in the accessibility tree (e.g., not 'hidden')
```
For figure, we're explicit that aria-labelledby, aria-describedby or aria-details prevent exposure of the details relation. However, we're not similarly explicit here for the reverse relation. Common sense will hopefully prevail, but I don't love relying on common sense for a spec. Is there something I'm missing that would otherwise make this explicit?

This is a similar situation to https://github.com/w3c/html-aam/pull/481/files#r1424943049. It seems we tend to use ARIA to express roles in a cross-platform manner wherever possible now, but we don't do that for relations, I guess largely because reverse relations and the details-roles attribute complicate things.

rahimabdi commented:
Closing, this PR has been migrated to the ARIA monorepo:
- https://github.com/w3c/aria/pull/2224
